packages/adblocker-extended-selectors/src/eval.ts

```diff
@@ -71,104 +105,297 @@ export function matches(element: Element, selector: AST): boolean {
+
+      const valueRegex =
+        valuePattern?.startsWith('/') && valuePattern.lastIndexOf('/') > 0
+          ? parseRegex(valuePattern)
+          : undefined;
+
+      if (namePattern.startsWith('/') && namePattern.lastIndexOf('/') > 0) {
+        // matching attribute name by regex
+        const regex = parseRegex(namePattern);
+        const matchingAttrs = [...element.attributes].filter((attr) => regex.test(attr.name));
+
+        // If no value pattern, return true if any attribute matches the name pattern
+        if (!valuePattern) {
+          return matchingAttrs.length > 0;
+        }
+
+        // Check if any of the matching attributes have the specified value
+        return matchingAttrs.some((attr) =>
+          valueRegex ? valueRegex.test(attr.value) : attr.value === valuePattern,
+        );
+      } else {
+        // matching attribute name by string
+        const value = element.getAttribute(namePattern);
+        // null means the attribute is not present
+        if (value === null) {
+          return false;
+        }
+
+        // early exit if no value pattern is provided
+        if (!valuePattern) {
+          return true;
+        }
+
+        return valueRegex ? valueRegex.test(value) : value === valuePattern;
+      }
     }
   }
 
   return false;
 }
 
-export function querySelectorAll(element: Element, selector: AST): Element[] {
-  const elements: Element[] = [];
+/**
+ * Tranposes the given element with a selector.
+ * @param element The subjective element
+ * @param selector A selector
+ * @returns An array with and without singular element; we may support transposing to multiple targets in the future
+ * but currently it's for the convenience to match return type.
+ */
+function transpose(element: Element, selector: AST): Element[] {
+  // The difference between `complex` handlers in `querySelectorAll` and `transpose` is nothing.
+  // However, it's also handled here to clarify and reduce complexity in the code flow.
+  // `querySelectorAll` is not used in `traverse` to avoid changing the subjective element unintentionally.
+  // Unlike this function, it assumes the subjective element to be guessed for other selectors.
+  // Also, we don't want to handle other selectors unless required as others are not *actually* transposable selectors.
+  // Since `traverse` assumes the subjective element to be the given element, we don't transpose to that.
+  // For an instance, type of `attribute` is to validate the subjective element has an `attribute`,
+  // not to select and transpose to the selector in a row.
+  // However, we want to do actual *transpose* not *match* with the selector type of `complex`
+  // as it could be mixed in the `compound` type selector and we want to move the target.
+  // e.g. This compound: `html>body` involes *transposing*, but not this compound: `html[lang*="en"]` (only *matching*).
+  // Therefore, this describes the behavior within the children nodes of the `compound` selector.
```

Comment:
`complex` is always transposing in short. Thinking `complex` branch in `querySelectorAll` as a fast exit helps to simplify the concept.